src/hotspot/share/utilities/debug.cpp

```diff
@@ -238,6 +238,33 @@ void report_vm_error(const char* file, int line, const char* error_msg)
   report_vm_error(file, line, error_msg, "%s", "");
 }
 
+static char _detail_msg[1024];
+
+static void print_error_for_unit_test(const char* message, const char* detail_fmt, va_list detail_args) {
```

Comment:
Trying to understand why you changed this code (beside factoring it out). We have:

- report_vm_out_of_memory -> message=NULL, detail set
- report_and_die from real crash -> message=NULL, detail=""
- report_fatal -> message="fatal error", detail set
- report_vm_error (assert, guarantee) -> message="assert (bla) failed", detail is set

so sometimes message is NULL, sometimes details are empty. What a mess :( Since you do gtests for all these cases now, you need to print the assert message correctly for them. You use _detail_msg as scratch buffer, since its already there. Later we print the detail message again into it. But I think that is fine.

I have a patch on the back burner somewhere simplifying these functions. Maybe we can improve matters. Then this function becomes easier too.

Reply:
I changed it to factor it out.  I think I got the combinations.